### ISSUE TITLE
`is_causal` parameter in torch.nn.TransformerEncoderLayer.forward does not work

### ISSUE THREAD

**Issue by JahooYoung** [pytorch/pytorch#96941]
### 🐛 Describe the bug

As the title says, `is_causal` parameter in torch.nn.TransformerEncoderLayer.forward does not work.

```python
import torch

d_model = 3
layer = torch.nn.TransformerEncoderLayer(d_model, 1, 6, batch_first=True)
layer.eval()
x = torch.randn(1, 5, d_model)
unmasked_output = layer(x)
is_causal_output = layer(x, is_causal=True)
mask = torch.nn.Transformer.generate_square_subsequent_mask(x.size(1))
masked_output = layer(x, src_mask=mask)

assert not torch.equal(unmasked_output, is_causal_output)
assert torch.equal(masked_output, is_causal_output)
```

### Versions

```
Collecting environment information...
PyTorch version: 2.0.0+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: Could not collect
CMake version: version 3.25.0
Libc version: glibc-2.35

Python version: 3.10.6 (main, Nov 14 2022, 16:10:14) [GCC 11.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-56-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090
GPU 2: NVIDIA GeForce RTX 3090
GPU 3: NVIDIA GeForce RTX 3090
GPU 4: NVIDIA GeForce RTX 3090
GPU 5: NVIDIA GeForce RTX 3090

Nvidia driver version: 525.60.11
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.5.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          80
On-line CPU(s) list:             0-79
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Xeon(R) Gold 6230 CPU @ 2.10GHz
CPU family:                      6
Model:                           85
Thread(s) per core:              2
Core(s) per socket:              20
Socket(s):                       2
Stepping:                        7
CPU max MHz:                     3900.0000
CPU min MHz:                     800.0000
BogoMIPS:                        4200.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts pku ospke avx512_vnni md_clear flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       1.3 MiB (40 instances)
L1i cache:                       1.3 MiB (40 instances)
L2 cache:                        40 MiB (40 instances)
L3 cache:                        55 MiB (2 instances)
NUMA node(s):                    2
NUMA node0 CPU(s):               0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42,44,46,48,50,52,54,56,58,60,62,64,66,68,70,72,74,76,78
NUMA node1 CPU(s):               1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39,41,43,45,47,49,51,53,55,57,59,61,63,65,67,69,71,73,75,77,79
Vulnerability Itlb multihit:     KVM: Mitigation: VMX disabled
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:          Mitigation; Enhanced IBRS
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Mitigation; TSX disabled

Versions of relevant libraries:
[pip3] numpy==1.24.2
[pip3] pytorch-lightning==2.0.0
[pip3] torch==2.0.0
[pip3] torch-tb-profiler==0.4.1
[pip3] torchdata==0.6.0
[pip3] torchmetrics==0.11.0
[pip3] torchtext==0.15.1
[pip3] torchvision==0.15.1
[pip3] triton==2.0.0
[conda] Could not collect
```

cc @jbschlosser @bhosmer @cpuhrsch @erichan1

**Comment by cpuhrsch** [comment]
cc @mikekgfb

**Comment by mikekgfb** [comment]
https://github.com/pytorch/pytorch/pull/97214 clarifies that is_causal is a hint that can be optionally provided.  

When the hint is provided, it may override use of the attention mask.  Undefined behavior occurs when is_causal is set and the mask is not a causal mask.

**Comment by thomasahle** [comment]
So basically, if I get you right, the way to make a causal model (or what some people might call a GPT/decoder only model) is this:
```
bs, seq, dim = 2, 3, 4
model = torch.nn.TransformerEncoder(encoder_layer=torch.nn.TransformerEncoderLayer(d_model=dim, nhead=2), num_layers=6)
out = model(torch.randn(seq, bs, dim), mask=torch.nn.Transformer.generate_square_subsequent_mask(seq), is_causal=True)
```
Though I'm still a bit confused about the purpose of setting is_causal, when you also use a causal mask.
And for some reason the code seems to run the same, even if the `seq` argument to `generate_square_subsequent_mask` is replaced with another value, like 0...

**Comment by mikekgfb** [comment]
> So basically, if I get you right, the way to make a causal model (or what some people might call a GPT/decoder only model) is this:
> 
> ```
> bs, seq, dim = 2, 3, 4
> model = torch.nn.TransformerEncoder(encoder_layer=torch.nn.TransformerEncoderLayer(d_model=dim, nhead=2), num_layers=6)
> out = model(torch.randn(seq, bs, dim), mask=torch.nn.Transformer.generate_square_subsequent_mask(seq), is_causal=True)
> ```
> 
> Though I'm still a bit confused about the purpose of setting is_causal, when you also use a causal mask. And for some reason the code seems to run the same, even if the `seq` argument to `generate_square_subsequent_mask` is replaced with another value, like 0...

is_causal is a hint that the mask is causal, and we don't actually have to use the mask if we have logic for hard-wired causal mask, such as in Flash Attention - when we process, we have the liberty to ignore the mask and just assume it's the causal mask.  

You don't need to supply is_causal, and evrything will work just fine (with small overhead for checking whether the mask is causal). Bad results will happen if is_causal is set when the supplied mask is not a causal mask.  It might work sometimes (when we look at is_causal), not at other times (when we look at the mask).

**Comment by patniemeyer** [comment]
The docs on this are super ambiguous: "is_causal: If specified, applies a causal mask as attention mask, and ignores attn_mask".  Yet it requires that you pass in a mask.  And what should you do if you have a causal mask that is not strictly causal but also ignores padding?  It is causal but it can't be ignored... so I guess you have to set causal to False?  This API really needs some attention : )
